### ISSUE TITLE
RFC: Support Rails 6

### ISSUE THREAD

**Issue by jarednorman** [solidusio/solidus#3053]
The first Rails 6 beta is out, so we can start discussing how we're going to support that. When the project moved from [Rails 4.x to Rails 5](https://github.com/solidusio/solidus/issues/1272), we cut a major version that introduced no new features and only upgraded the codebase for compatibility. Depending on the extent of the changes in Rails 6, something that big might not be necessary.

I'm opening this issue to gather comments/discussion on Rails 6 support for Solidus.

**Comment by mdesantis** [comment]
One thought about Rails 6 and ActionText: are there some places in Solidus where we'd like to use it? Any places where we write rich text?

**Comment by kennyadsl** [comment]
@mdesantis [solidus_editor](https://github.com/solidusio-contrib/solidus_editor) is what is used for this kind of things most of the times, I think.

**Comment by ericsaupe** [comment]
This one too [solidus_static_content](https://github.com/solidusio-contrib/solidus_static_content). I fired up a new Rails 6 app with ActionText and it worked pretty well. Definitely could make those extensions cleaner. It works with ActiveStorage so it might be something that benefits from #2974

**Comment by jarednorman** [comment]
I've been live streaming myself working on this. Here's the [YouTube playlist of past streams](https://www.youtube.com/playlist?list=PLDCxp49uNb9wej1n94oaR_bieiqzp59RO).

The biggest issue I've run into so far is that there are some really annoying incompatibilities with newer versions of CanCanCan. I've been running against the branch that has Rails 6 support, and it's caused a whole bunch of unrelated issues.

I'll continue working away on that. I haven't opened a progress PR because I'm waiting on releases on a bunch of the gems that will need to be bumped to support Rails 6. Currently I'm just pointing at various GitHub branches that contain the tentative work on Rails 6 support for those gems.

I've got core passing, and backend/frontend almost completely passing if I comment out the stuff in the permission sets that are causing the CanCanCan headaches. API is full of failing specs and I haven't dug into that, but I imagine it's at least partially related to the authorization issues.

**Comment by coorasse** [comment]
The pull request #3148 tackles the issues with cancancan 3.0 by reorganising abilities in a proper way.
The pull request #3174 overrides the merge method of cancancan 3.0 to keep the old behaviour and be able to proceed with the upgrade to Rails 6.0

**Comment by aldesantis** [comment]
I am working on this in #3236. The work is basically done on the Solidus side, but there are a few dependencies that are not yet ready for Rails 6, so we're waiting on those!

**Comment by ericsaupe** [comment]
Rails 6 has now officially been released

https://weblog.rubyonrails.org/2019/8/15/Rails-6-0-final-release/

https://rubygems.org/gems/rails/versions/6.0.0

**Comment by aldesantis** [comment]
https://github.com/solidusio/solidus/pull/3236 is ready to merge.

Tasks that are not blocking and will be done in separate PRs:

- [ ] Replace `update_attributes` with `update`
- [ ] Adopt RSpec 4.x stable when it comes out

**Comment by cedum** [comment]
Just discovered that Rails 6' sqlite3 adapter [got bumped to `v1.4`](https://github.com/rails/rails/blob/v6.0.0/activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb#L13).
This causes the same kind of conflicts described and fixed here [PR #3088 (Lock sqlite3 version to 1.3)](https://github.com/solidusio/solidus/pull/3088) for Rails 5.2.
Actually, on Solidus, sqlite3 is fixed to `~> v1.3.6` and on Rails6 to `~> v1.4` (not `~> v1.4.0`)
Any ideas? 💡

**Comment by tvdeyen** [comment]
This only effects the sandbox, correct?

**Comment by kennyadsl** [comment]
It also effects running bundle install in development with sqlite and rails 6. I'm taking a look and have a PR almost ready for that.

**Comment by tvdeyen** [comment]
Awesome. Thanks. Shouldn’t the solution be as easy as removing the version constraint. So, that Bundler resolves the correct version for us dependent on which Rails Version we use?

**Comment by kennyadsl** [comment]
I expected an issue actually, since Rails does not have any direct sqlite dependency it will always use the latest one. In fact, running 

```
RAILS_VERSION='~> 5.2.0' bundle install
```

generates a Gemfile.lock with Rails `5.2.3` and sqlite3 `1.4.1`, which should not allow completing the bundle. Maybe I'm missing something or latest `5.2.x` Rails versions are now compabile with sqlite `1.4.1`. Still investigating...

**Comment by cedum** [comment]
@kennyadsl the sqlite3 version got "relaxed" since rails `v5.2.3`:
https://github.com/rails/rails/blob/v5.2.2/activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb
`v5.2.2` instead contains the stricter one:
https://github.com/rails/rails/blob/v5.2.2/activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb

**Comment by kennyadsl** [comment]
@cedum that explains what's happening, thanks!

**Comment by aldesantis** [comment]
@kennyadsl can we close this?

**Comment by kennyadsl** [comment]
I think we "just" need to update all extensions to work with Zeitwerk to have a fully working Rails 6 support, but yes, we should probably close this one and create a new issue? What do you think?

**Comment by aldesantis** [comment]
@kennyadsl we should be good to go since https://github.com/spree/deface/pull/202 was merged.

**Comment by spaghetticode** [comment]
Closing, since Solidus is compatible with Rails 6 since version 2.10.
